### ISSUE TITLE
[BUG] Opening start menu blocks rendering and causes high cpu usage

### ISSUE THREAD

**Issue by canwail** [linuxmint/cinnamon#11103]
```
 * Cinnamon version: 5.4.10
 * Distribution: Manjaro
 * Graphics : Nvidia GTX 1060 6GB / Driver Version: 515.65.01
 * 32 or 64 bit: 64bit
 * Attach ~/.xsession-errors, or /var/log/syslog: - 
 ```

### Issue
High CPU usage when opening the start menu and blocking the rendering of other applications

https://user-images.githubusercontent.com/50383630/185905244-050929e6-dbef-4cf1-bab3-8e2111072ca9.mp4

In the video it can be seen that the problem occurs even on a freshly added panel (no additional applets added).
Additionally it can be seen that other applications are blocked from rendering (e.g. The DVD Video temporarily stops moving) while the start menu is loading. The "Disk write" tab in the system monitor shows that each time the start menu is opened, data is written to disk for some reason. I was unable to find what or where cinnamon is writing to. 

The "loading time" it takes for the start menu to show up gets worse with time, after using the system for 8-10 hours the loading time is substantial. Running `cinnamon --replace` manually in the terminal fixes the issue, meaning its as fast as after a fresh reboot, but it gets slower and slower with time again.

### Steps to reproduce
Spam the Windows Key or click the start button repeatedly

### Expected behaviour
Not-as-high cpu usage and not blocking the rendering other applications

### Other information
If someone can point me to a guide on how to compile and install cinnamon with debug symbols I could try profiling it to see where the cpu time is spent, I tried building this repository, but even after running `ninja install` I could not find a cinnamon binary on my system apart from the one installed via package manager.

**Comment by mtwebster** [comment]
Use `meson yourbuilddir --prefix=/usr` - cinnamon doesn't really work in any non-standard location (like /usr/local).

I can't reproduce this - I do get small blips when opening the menu repeatedly while playing that video, but nothing like you see, and it doesn't get worse. CPU will momentary spike when opening the menu but that's pretty normal - everything is on the same process here.

Just our of curiosity, does this happen using the default Cinnamon theme?

**Comment by canwail** [comment]
I successfully managed to build and install cinnamon and muffin with debug symbols, however running `perf record` shows that the offending module is "?? [??]". Im no expert on perf, but since it doesnt show any information (not even the caller) on it made me think that this could be system calls:

```
$ sudo strace -p $(pidof cinnamon) -C -o cinnamon.strace
$ tail -n27 cinnamon.strace
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 98.77    7.474428           5   1265842           write
  0.55    0.041610           7      5552       658 newfstatat
  0.32    0.023990           4      5434      4756 ioctl
  0.16    0.011852           2      5491           getpid
  0.08    0.006036           1      3181      2497 recvmsg
  0.05    0.003822           2      1580           poll
  0.03    0.002426           4       547        23 futex
  0.02    0.001313           2       605           writev
  0.01    0.001096           1       653           mprotect
  0.00    0.000277           1       247           read
  0.00    0.000162           0       224           sched_yield
  0.00    0.000069           1        57           recvfrom
  0.00    0.000058           3        15           mmap
  0.00    0.000009           1         5           clone
  0.00    0.000005           5         1           restart_syscall
  0.00    0.000002           2         1           madvise
  0.00    0.000002           2         1           getrandom
  0.00    0.000000           0         3           close
  0.00    0.000000           0        10           rt_sigprocmask
  0.00    0.000000           0        12           readlink
  0.00    0.000000           0         2           getrusage
  0.00    0.000000           0         3           timerfd_create
  0.00    0.000000           0         3           timerfd_settime
------ ----------- ----------- --------- --------- ----------------
100.00    7.567157           5   1289469      7934 total

```

I tried getting more detailed info with strace-analyzer, but it wouldnt give me any output. Ill see wether i can get more useful information (for example which paths/files are most frequently written to) any other way. Till then, here is the raw strace output I recorded (about 10-15s worth of data, 60mb uncompressed): [cinnamon.strace.gz](https://github.com/linuxmint/cinnamon/files/9396165/cinnamon.strace.gz)

**Comment by canwail** [comment]
I found that whenever I the hang occurs, `write` is called with `fd=2`, which corresponds to `.xsession-errors`. By watching the file with `tail -F` I found that just simply opening and closing the menu once with the start button writes about 5500 lines to said .xsession-errors. Here a small excerpt:

```
(cinnamon:1432): Gjs-CRITICAL **: 20:30:24.419: Object St.Bin (0x561420317fc0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x56141e6ad170 ==
#0   7ffc845c2460 b   /usr/share/cinnamon/js/ui/layout.js:715 (231a44a89240 @ 310)
#1   7ffc845c2520 b   self-hosted:844 (677ff0af9c0 @ 398)

(cinnamon:1432): Gjs-CRITICAL **: 20:30:24.419: Object St.Bin (0x561420317fc0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x56141e6ad170 ==
#0   7ffc845c2460 b   /usr/share/cinnamon/js/ui/layout.js:716 (231a44a89240 @ 575)
#1   7ffc845c2520 b   self-hosted:844 (677ff0af9c0 @ 398)

(cinnamon:1432): Gjs-CRITICAL **: 20:30:24.419: Object St.Bin (0x561420317fc0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x56141e6ad170 ==
#0   7ffc845c2460 b   /usr/share/cinnamon/js/ui/layout.js:732 (231a44a89240 @ 1098)
#1   7ffc845c2520 b   self-hosted:844 (677ff0af9c0 @ 398)
```
Even if this hints at some underlaying problem, opening a simple start menu should not cause that many lines to be written and even if, it should not block the UI. 

I could be reading this incorrectly but looking at my trace it seems that `write` is often called for one character at a time with no buffering, which would probably explain the high cpu usage and the high number of `write` calls (1.2 Million in 10 seconds):
```
362. write(2, "\n", 1)                       = 1
363. write(2, "=", 1)                        = 1
364. write(2, "= Stack trace for context 0x5614"..., 44) = 44
365. write(2, "#", 1)                        = 1
366. write(2, "0", 1)                        = 1
367. write(2, " ", 1)                        = 1
368. write(2, " ", 1)                        = 1
369. write(2, " ", 1)                        = 1
370. write(2, "7ffc845c0560", 12)            = 12
371. write(2, " ", 1)                        = 1
372. write(2, "b", 1)                        = 1
373. write(2, " ", 1)                        = 1
374. write(2, " ", 1)                        = 1
375. write(2, " ", 1)                        = 1
376. write(2, "/usr/share/cinnamon/js/ui/layout"..., 35) = 35
377. write(2, ":", 1)                        = 1
378. write(2, "647", 3)                      = 3
379. write(2, " ", 1)                        = 1
380. write(2, "(", 1)                        = 1
381. write(2, "2", 1)                        = 1
382. write(2, "31a44a89470", 11)             = 11
383. write(2, " ", 1)                        = 1
384. write(2, "@", 1)                        = 1
385. write(2, " ", 1)                        = 1
386. write(2, "174", 3)                      = 3
387. write(2, ")", 1)                        = 1
388. write(2, "\n", 1)                       = 1
389. write(2, "#", 1)                        = 1
390. write(2, "1", 1)                        = 1
```

**Comment by mtwebster** [comment]
The warnings are pointing out things that were probably pre-existing issues. What's new is their getting logged. It's a good thing honestly - they should be fixed (and they are), but can be challenging to track down.

Is this the default cinnamon menu? That's definitely abnormal (for any of my machines, the menu applet produces no logging). I'm not sure about any 3rd-party menus.

**Comment by canwail** [comment]
In that case I used a 3rd party one, original one was with the default one. I will check later to verify whether both produce the same error lines or different ones. However even if those lines are deemed necessary, writing logs should not block the UI thread. Even if a third party applet were to blame here I think this is something that should be adressed, since the logs *are* coming from cinnamon, and not the applet itself.

**Comment by canwail** [comment]
Yes, happens with default menu too. And not just that, happens when I interact with any applet on the panel (Calendar, Network Manager) or just ALT+TAB. Even running the `tail -F .xsession-errors` command in terminal and just clicking another window (switching focus) and going back to CTRL+C the tail command produces 1200 lines. The output is always the same as above: `/usr/share/cinnamon/js/ui/layout.js:` 715, 716 and 732.

https://github.com/linuxmint/cinnamon/blob/8d8180f1ec6ff2e1cd1a2d8ff446e3f575215613/js/ui/layout.js#L715-L732

**Comment by canwail** [comment]
I tried guarding the access to actor and thought that it had worked for a while apparently not. I noticed that when opening a context menu different lines get logged, but still related to `.actor` accesses:
```
(cinnamon:1444): Gjs-CRITICAL **: 14:11:16.633: Object St.Bin (0x55cbab851bd0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x55cbab0bb1a0 ==
#0   7fff33c918a0 b   /usr/share/cinnamon/js/ui/layout.js:648 (19f02bc89470 @ 433)
#1   7fff33c91970 b   /usr/share/cinnamon/js/ui/layout.js:656 (19f02bc89420 @ 30)
#2   7fff33c91a50 b   /usr/share/cinnamon/js/ui/layout.js:589 (19f02bc89650 @ 476)
#3   55cbab869a38 i   /usr/share/cinnamon/js/ui/layout.js:695 (19f02bc89290 @ 130)
#4   55cbab869968 i   self-hosted:844 (2e065e3af9c0 @ 423)

(cinnamon:1444): Gjs-CRITICAL **: 14:11:16.633: Object St.Bin (0x55cbab851bd0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x55cbab0bb1a0 ==
#0   7fff33c91a50 b   /usr/share/cinnamon/js/ui/layout.js:593 (19f02bc89650 @ 557)
#1   55cbab869a38 i   /usr/share/cinnamon/js/ui/layout.js:695 (19f02bc89290 @ 130)
#2   55cbab869968 i   self-hosted:844 (2e065e3af9c0 @ 423)

(cinnamon:1444): Gjs-CRITICAL **: 14:11:16.634: Object St.Bin (0x55cbadf7e9f0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x55cbab0bb1a0 ==
#0   7fff33c918a0 b   /usr/share/cinnamon/js/ui/layout.js:647 (19f02bc89470 @ 174)
#1   7fff33c91970 b   /usr/share/cinnamon/js/ui/layout.js:656 (19f02bc89420 @ 30)
#2   7fff33c91a50 b   /usr/share/cinnamon/js/ui/layout.js:589 (19f02bc89650 @ 476)
#3   55cbab869a38 i   /usr/share/cinnamon/js/ui/layout.js:695 (19f02bc89290 @ 130)
#4   55cbab869968 i   self-hosted:844 (2e065e3af9c0 @ 423)

```

**Comment by canwail** [comment]
I tried building cinnamon with c7110ea25b03226eb974cb24d5b390e7ce61cdec which I assume was meant to address this issue but unfortunately to no avail: 

https://user-images.githubusercontent.com/50383630/187443854-a0ab5447-07fe-4736-866e-bc16a7ef00f3.mp4

The video shows me opening and closing the context menu, switching focus (0:02) and alt+tabbing to produce log lines.

**Comment by mtwebster** [comment]
Can you try to reproduce this with a new user account?

I can't reproduce this on Manjaro either

**Comment by canwail** [comment]
I tried it with a new live ISO of Linux Mint 21 and the fastest way to reproduce is:

1. `tail -F .xsession-errors` (to see when log spamming starts)
2. Right click on panel -> Applets -> Download -> "CobiWindowList" -> Install. (Minor logspam here, but not the one I mean)
3. Go to Manage in the Applets window and add "CobiWindowList" to panel.
4. Open and close Rythmbox (probably works with others too, but thats the first i tried and it comes with Mint)

As soon as Rythmbox is closed, `.xsession-errors` should start to be spammed with logs when switching focus, opening and closing start menu, opening context menu etc. Removing and uninstalling the applet will not stop the log spam.  Im not sure its only this applet in particular, but its the first I tried that works reliably.
